madrid/bicimad.py

```diff
@@ -0,0 +1,144 @@
+            timeout=timeout
+        )
+
+        if response.status_code != 200:
+            raise requests.HTTPError(f'The status code of request to {url} was {response.status_code}')
+
+        response_json = response.json()
+
+        return response_json
+
+    @shout
+    def _load_access_token(self):
+        """
+        Loads working access token into self
+        """
+        try:
+            self.access_token = open('access_token.txt', 'r').read()
+        except FileNotFoundError:
+            self.access_token = None
+
+        if (self.access_token is None) or (not self._is_access_token_alive()):
+            self._get_access_token()
+
+    @shout
+    def _get_access_token(self):
+        """
+        Calls API to get new access token
+        """
+        print('Getting access token...')
+
+        response_json = self._get(url=self.url_login, headers=auth_bicimad)
+
+        if (code := response_json.get('code')) == '00':
+            self.access_token = response_json.get('data')[0].get('accessToken')
+            self._save_access_token()
+
+        elif code == '01':
+            print('Logging out...')
+            self.access_token = response_json.get('description').split()[1]
+            self._save_access_token()
+            self._logout()
+            self._get_access_token()
+
+    @shout
+    def _logout(self):
+        """
+        Calls API to logout
+        """
+        self._get(url=self.url_logout, headers={'accessToken': self.access_token})
+
+    @shout
+    def _is_access_token_alive(self):
+        """
+        Checks if current access token works
+        """
+        if self.access_token is None:
+            return False
+
+        response_json = self._get(url=self.url_who_am_i, headers={'accessToken': self.access_token})
+
+        if response_json.get('code') != '02':
+            return False
+        else:
+            return True
+
+    @shout
+    def _save_access_token(self):
+        """)
+        Saves access token to file
+        """
+        open('access_token.txt', 'w+').write(self.access_token)
+
+    @shout
+    def _get_stations_raw_info(self):
+        """
+        Calls API and gets all stations' status
+        """
+        response_json = self._get(url=self.url_get_all_stations_info, headers={'accessToken': self.access_token})
+
+        if response_json.get('code') != '00':
+            print('Invalid token... lets get a new one')
+            self._get_access_token()
+
+            return self._get_stations_raw_info()
+
+        n_stations = int(response_json.get('description')[:3])
+        datetime = response_json.get('datetime')
+        data = response_json.get('data')
+
+        return n_stations, datetime, data
+
+    @shout
+    def get_stations_dataframe(self):
+        n_stations, datetime, data = self._get_stations_raw_info()
+
+        df = pd.DataFrame(data)[self.stations_keys]
+
+        df['datetime'] = pd.to_datetime(datetime)
+
+        return df
```

Comment:
This is just an observation but do we really need to convert the JSON to a `pd.DataFrame`? I feel that's a step that could be avoided and no need to import the `pandas` dependency. In the end, we're just gonna push the JSON in someway to the database. I guess `pandas` will be needed when we export it from database to do some type of analysis.

Reply:
I totally agree. Let me change it.